src/core/qgspallabeling.cpp

```diff
@@ -1186,11 +1221,29 @@ void QgsPalLayerSettings::calculateLabelSize( const QFontMetricsF *fm, const QSt
 
 void QgsPalLayerSettings::registerFeature( const QgsFeature &f, QgsRenderContext &context, QgsLabelFeature **labelFeature, QgsGeometry obstacleGeometry )
 {
+  QgsFeature feature = f;
+  if ( geometryGeneratorEnabled )
+  {
+    if ( !mGeometryGeneratorExpression.isValid() )
+    {
+      mGeometryGeneratorExpression = QgsExpression( geometryGenerator );
+      mGeometryGeneratorExpression.prepare( &context.expressionContext() );
+      if ( mGeometryGeneratorExpression.hasParserError() )
+        QgsMessageLog::logMessage( QObject::tr( "Labeling" ), mGeometryGeneratorExpression.parserErrorString() );
+    }
+    context.expressionContext().setFeature( feature );
+    const QgsGeometry geometry = mGeometryGeneratorExpression.evaluate( &context.expressionContext() ).value<QgsGeometry>();
+    if ( mGeometryGeneratorExpression.hasEvalError() )
+      QgsMessageLog::logMessage( QObject::tr( "Labeling" ), mGeometryGeneratorExpression.evalErrorString() );
+
+    feature.setGeometry( geometry );
```

Comment:
I'm unsure about the impact of changing the feature's geometry like this. I believe it will mess with expression based labels (e.g. those showing the feature's area/length) and also mess with geometry based data defined settings. I think it's safest to keep the geometry-for-label-position and original feature separate from here.

Reply:
It's a copy of the feature with is modified, the original feature stays untouched.
I'll switch the data defined and obstacle code back to the original feature, does that work?